service/csvService.go

```diff
@@ -6,133 +6,219 @@ import (
+	"github.com/labstack/gommon/log"
 )
 
+// CsvService is used as an interface for this service.
 type CsvService interface {
-	FilterById(string) ([]Doc, error)
-	DownloadCsvData() error
+	FilterByID(string) ([]model.Doc, error)
+	DownloadCsvData(string) error
+	SearchByConditions(map[string]string) ([]model.Doc, error)
 }
 
+// CsvServiceImpl is used as the implementation of this service.
 type CsvServiceImpl struct{}
 
-type SearchResponse struct {
-	Doc      []Doc `json:"docs"`
-	NumFound int   `json:"numFound"`
-	Start    int   `json:"start"`
+// DownloadCsvData downloads information from external API.
+func (c CsvServiceImpl) DownloadCsvData(topic string) error {
+	return getAndSaveData(topic)
 }
 
-type Doc struct {
-	Key       string `json:"key"`
-	Title     string `json:"title"`
-	Type      string `json:"type"`
-	Published string `json:"first_published_year"`
-}
+// FilterByID reads records from a CSV and returns the filtered ones
+func (c CsvServiceImpl) SearchByConditions(conditions map[string]string) ([]model.Doc, error) {
+	file, err := os.Open(getConfigService().GetConfig().CSV.FileName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file: %v", err)
+	}
+	reader := csv.NewReader(file)
+	records, errReader := reader.ReadAll()
+	if errReader != nil {
+		return nil, errReader
+	}
+
+	var recordsStruct []model.Doc
+	for _, record := range records {
 
-var CSVFile = "/home/omar/Desktop/saved.csv"
+		docRecord := toDoc(record)
+		fields := structs.Fields(docRecord)
 
-// func (c )LoadCsvData
-func (c CsvServiceImpl) DownloadCsvData() error {
-	getAndSaveData()
-	return nil
+		if ( filterRecord(fields, conditions) ) {
+			recordsStruct = append(recordsStruct, toDoc(record))
+		}
+
+	}
+
+	if len(records) == 0 {
+		log.Info("The file is empty")
+		return nil, errors.New("The file is empty")
+	}
+	defer file.Close()
+
+	return recordsStruct, nil
 }
 
-func (c CsvServiceImpl) FilterById(id string) ([]Doc, error) {
-	file, err := os.Open(CSVFile)
+// FilterByID reads records from a CSV and returns the filtered ones
+func (c CsvServiceImpl) FilterByID(id string) ([]model.Doc, error) {
+	file, err := os.Open(getConfigService().GetConfig().CSV.FileName)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to open file: %v", err)
 	}
 	reader := csv.NewReader(file)
 	records, errReader := reader.ReadAll()
 	if errReader != nil {
 		return nil, errReader
 	}
 
-	var recordsStruct []Doc
+	var recordsStruct []model.Doc
+	conditions := make(map[string] string)
+	conditions["Key"] = id
 
 	for _, record := range records {
-		fmt.Println(record)
 
-		doc := Doc{
-			Key:       record[0],
-			Title:     record[1],
-			Type:      record[2],
-			Published: record[3],
-		}
-		if doc.Key == id {
-			recordsStruct = append(recordsStruct, doc)
+		// if record[0] != id {
+			// continue
+		// }
+		docRecord := toDoc(record)
+		fields := structs.Fields(docRecord)
+
+		if ( filterRecord(fields, conditions) ) {
```

Comment:
```suggestion
		if filterRecord(fields, conditions) {
```

Use `gofmt` to format documents.

Reply:
I was using EMACS I couldn't figure out how to configure it, so in VS Code is working out of the box, I have those extensions now.